### PR TITLE
GLSL validation updates

### PR DESCRIPTION
**Pull request by bernardkwok**
- Add in binding of lights and IBL properly for shaders
-- Cleanup texture handling and texture reuse
- Add in transparency blending if shader is transparency
- Add in switch to use different geometry and camera for shaders
- Add in standard surface as a graph for testing.
- Fix memory leak on image load from disk
- Fix transparency detection to consider color4/vec4 inputs to output graph as being possibly transparent.
- Add in new texture with a mask to allow checking.
- Change OSL color4 and vec4 output to save pre-multiplied alpha values.

### CODE REVIEW

File: source/MaterialXRender/ShaderValidators/Glsl/GlslProgram.cpp [hunk 1]
```diff
@@ -560,50 +564,40 @@ void GlslProgram::createDummyTexture(ImageHandlerPtr imageHandler)
...
-        throw ExceptionShaderValidationError(errorType, errors);
-    }
+        // Use next available slot
+        if (_maxImageUnits < 0)
+        {
+            glGetIntegerv(GL_MAX_COMBINED_TEXTURE_IMAGE_UNITS, &_maxImageUnits);
+        }
+        if (_activeTextureUnit >= _maxImageUnits)
+        {
+            return textureBound;
```
Comment (feldstj) [inline source/MaterialXRender/ShaderValidators/Glsl/GlslProgram.cpp]: Is it more clear to just return false here?

File: source/MaterialXRender/ShaderValidators/Glsl/GlslProgram.cpp [hunk 2]
```diff
@@ -300,13 +306,26 @@ void GlslProgram::bindInputs(ViewHandlerPtr viewHandler,
...
+        glDisable(GL_BLEND);
+    }
 }
 
 void GlslProgram::unbindInputs()
 {
     unbindTextures();
     unbindGeometry();
+
+    glDisable(GL_BLEND);
```
Comment (feldstj) [inline source/MaterialXRender/ShaderValidators/Glsl/GlslProgram.cpp]: Minor: Should you only do this if _hwShader->hasTransparency()? I mean you can do it either way, but I think the point is you are only doing it for this case, no?

File: source/MaterialXRender/ShaderValidators/Glsl/GlslProgram.h [hunk 3]
```diff
@@ -266,8 +270,13 @@ class GlslProgram
...
 
-    /// Program textures
-    std::vector<unsigned int> _programTextures;
+    /// Program texture map
+    std::unordered_map<std::string, unsigned int> _programTextures;
+
+    /// Maximum image units
+    int _maxImageUnits;
+    /// Active texture units used
+    int _activeTextureUnit;
```
Comment (feldstj) [inline source/MaterialXRender/ShaderValidators/Glsl/GlslProgram.h]: Just a suggestion: Maybe rename this to something like _textureUnitsInUse; ?

File: source/MaterialXRender/ShaderValidators/Glsl/GlslProgram.cpp [hunk 4]
```diff
@@ -614,33 +612,84 @@ void GlslProgram::bindTextures(ImageHandlerPtr imageHandler)
...
                     glBindTexture(GL_TEXTURE_2D, newTexture);
                     glTexImage2D(GL_TEXTURE_2D, 0, GL_RGBA, width, height,
                         0, (channelCount == 4 ? GL_RGBA : GL_RGB), GL_FLOAT, buffer);
-                    glGenerateMipmap(GL_TEXTURE_2D);
+                    if (generateMipMaps)
+                    {
+                        glGenerateMipmap(GL_TEXTURE_2D);
+                    }
+
+                    free(buffer);
```
Comment (bernardkwok) [inline source/MaterialXRender/ShaderValidators/Glsl/GlslProgram.cpp]: This was a memory leak.

**Pull request merged**